smithy-model/src/main/java/software/amazon/smithy/model/shapes/EntityShape.java

```diff
@@ -25,12 +27,39 @@
 public abstract class EntityShape extends Shape {
 
     private final Set<ShapeId> resources;
+    private final Set<ShapeId> introducedResources;
     private final Set<ShapeId> operations;
+    private final Set<ShapeId> introducedOperations;
 
     EntityShape(Builder<?, ?> builder) {
         super(builder, false);
-        resources = builder.resources.copy();
-        operations = builder.operations.copy();
+
+        if (getMixins().isEmpty()) {
+            resources = builder.resources.copy();
+            introducedResources = resources;
+            operations = builder.operations.copy();
+            introducedOperations = operations;
+        } else {
+            Set<ShapeId> computedResources = new TreeSet<>();
+            Set<ShapeId> computedOperations = new TreeSet<>();
+
+            for (Shape shape : builder.getMixins().values()) {
+                if (shape instanceof EntityShape) {
```

Comment:
I think you’re doing checks during the loading process to make sure mixins are coherent right? If so, it would be fine to throw here and do some up front validation.

Reply:
Yeah it ensures the shape types match before this is run.